index.d.ts

```diff
@@ -0,0 +1,104 @@
+// TypeScript Version: 3.7
+
+import {Test} from 'unist-util-is'
+import * as unist from 'unist'
+
+export {}
+
+type Null = undefined | null
+
+declare namespace messageControl {
+  export {}
+  /** a comment marker */
+  export interface Marker<N extends unist.Node> {
+    /** name of marker */
+    name: string
+    /** value after name */
+    attributes: string
+    /** parsed attributes */
+    parameters: Record<string, unknown>
+    /** reference to given node */
+    node: N
+  }
+
+  /** parse a possible comment marker node to a Marker */
+  export interface MarkerFn<N extends unist.Node> {
+    (node: N): Marker<N> | Null
+  }
+
+  interface MessageControlOptionsWithReset<T extends unist.Node>
+    extends BaseMessageControlOptions<T> {
+    // must be explicitly set to `true` because reset only defaults
+    // to `false`.
+    reset: true
+    disable?: string[]
+    enable?: undefined
+  }
+
+  interface MessageControlOptionsWithoutReset<T extends unist.Node>
+    extends BaseMessageControlOptions<T> {
+    reset?: false | Null
+
+    enable?: string[] | Null
+    disable?: Null
```

Comment:
@ChristianMurphy Do you think we need to care about `null`?
AFAIK, we've never allow `null` for option prop in type definitions because we're using `Partial`. 

```ts
type Partial<T> = {
    [P in keyof T]?: T[P];
};
```

Reply:
I don't think we need `null` as part of the type.
Also with `Partial` the `?` annotation is not needed.
In addition in the docs for this plugin:
> options.disable
> List of ruleIds to turn on if reset: false (Array.<string>, optional).

I think `disable` should have a `string[]` included in the typing

Reply:
Partial makes all the fields optional, but the documentation states that only some of the fields are optional. Null is used here for the same reason as https://github.com/DefinitelyTyped/DefinitelyTyped/pull/41631 -- the use of `null` vs `undefined` seems to be interchangeable within the unist ecosystem. In standard typescript null and undefined are interchangeable, but it seems that now it's considered best practice to separate them. see: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#--strictnullchecks

Reply:
(in this case, the DefinitelyTyped type definitions were actually *rejecting* the mdast examples because they didn't allow `null`)

Reply:
AFAIK, handling `null` as `undefined` is a handy casual convention but not a spec in unified ecosystem. So, unless our specs said explicitly that `null` is allowed, I would go for using `?` only without caring `null`.

Reply:
@wooorm How do you think? Should we explicitly allow `null` type for optional props or ignore `null` type like how we've done before?

Reply:
I think in programming it is wise to stay close to the language: JavaScript is loose, so when in JS, embrace that and be loose as well. E.g., if expecting a string and a number is passed in, that number should be serialised to string. Same with `null` and `undefined` (I believe this is called “nully”) when something is optional.
Now, there are two competing viewpoints here: a) TS is strict; b) TS should be close to the implementation.
I don’t know which one should have more weight! 🤷‍♂️

Reply:
I'd lean towards more strict.

If we go in the direction of supporting everything javascript does, we may also need to add all the falsy values to the typing: `false`, `0`, `NaN`, `0n`, and `''`.
While that technically may be supported, I'm not sure that's something we want to encourage adopters to leverage.

Reply:
The issue here is while I appreciate the strictness, I feel strongly that I don't want to type anything that would break existing code in the unified ecosystem, as noted on my previous comment. The DefinitelyTyped standard is to say if the type could be potentially `null` or `undefined`, and I think where it was never strictly specified canonically previously I'd like to call out that `null` and `undefined` should be expected.

If we don't do this, then anyone without the relevant TypeScript flag is going to get errors when `null` is encountered instead of `undefined` or vice-versa ... the TypeScript system will make sure you do (!field) as a guard instead of `field !== undefined` if we specify both `null` and `undefined`.

On `false`, `0`, `NaN` etc, I think there's a point where it gets unreasonable – I don't think many would say that these can / should / are used interchangeably with undefined in the unified ecosystem and as such I don't think it makes sense to support it.

Reply:
> I feel strongly that I don't want to type anything that would break existing code in the unified ecosystem

Adding typings is considered a breaking change in the Unified ecosystem.
The main reason, is that the typings we provide target TypeScript 3+, and some adopters use TypeScript v1 and v2.

Given that it is already a potentially breaking change, we include any other potentially breaking changes that may improve clarity at the same time.

> The DefinitelyTyped standard is to say if the type could be potentially `null` or `undefined`

Do you have a citation for this?
This is not included in the best practice guide https://definitelytyped.org/guides/best-practices.html or the contributing guide https://definitelytyped.org/guides/contributing.html

> If we don't do this, then anyone without the relevant TypeScript flag is going to get errors when null is encountered instead of undefined or vice-versa ... the TypeScript system will make sure you do `(!field)` as a guard instead of `field !== undefined` if we specify both `null` and `undefined`.

Yes, that is what a type checker does.

Reply:
@ChristianMurphy it's very clearly stated here

https://github.com/DefinitelyTyped/DefinitelyTyped#some-packages-have-no-tslintjson-and-some-tsconfigjson-are-missing-noimplicitany-true-noimplicitthis-true-or-strictnullchecks-true 

> Yes, that is what a type checker does.
i ... don't really understand your response. I know what a typechecker does. My point is that if we **do not** specify `null`, typescript may allow `(field !== undefined)`. Since this does not exclude `null`, runtime bugs can and will happen because `null` fields will be considered 'defined'.

Reply:
> https://github.com/DefinitelyTyped/DefinitelyTyped#some-packages-have-no-tslintjson-and-some-tsconfigjson-are-missing-noimplicitany-true-noimplicitthis-true-or-strictnullchecks-true

Not it does not.
This states that three compiler options should be enabled.
1. These compiler options are already included in the unified typings, as part of `"strict": true`
2. It states this should be done for typing themselves, it does not apply to downstream adopters. Meaning if adopters want a looser interpretation of `null` and `undefined` the unified API the can have it.
3. This speaks only to the compiler options, not to whether project should explicitly type with `null`, `undefined` or both.

> i ... don't really understand your response. I know what a typechecker does.

It means you stated what a type checker will do in a specific circumstance, _by itself_, this information has no bearing on the conversation.
It would have bearing if:

- choosing just undefined would break `unified-message-control` itself
- choosing just `undefined` would mislead adopters
- choosing just `undefined` would change the stated API

With my current knowledge, I don't believe any of those three is true in this case.
This code supports `undefined`, stating so would not mislead adopters, and `null` is only explicitly supported as part of the API for the `unist-util-is` test.

> My point is that if we do not specify null, typescript may allow `(field !== undefined)`. Since this does not exclude null, runtime bugs can and will happen because null fields will be considered 'defined'.

Yes, but where would that happen?
All the potentially nullable values are going into `unified-message-controller`, not coming out.
What happens inside `unified-message-controller` is an implementation detail, only callbacks and return values would have to potential to impact adopters.